unliftio/test/UnliftIO/ExceptionSpec.hs

```diff
@@ -79,6 +79,27 @@ spec = do
       result <- withWrappedAsyncExceptionThrown $ \m -> trySyncOrAsync (void m)
       first fromExceptionUnwrap result `shouldBe` Left (Just Exception1)
 
+  describe "withException" $ do
+    it "should work when withAsync is in the handler" $ do
+      let
+        action =
+          error "oops"
+            `onException` do
+              let
+                timerAction n = do
+                  threadDelay 1000000
+                  when (n < 10) $ do
+                    timerAction (n + 1)
+              withAsync (timerAction 0) $ \a -> do
+                cancel a
+      eresult <-
+        race
```

Comment:
This `race` call *also* is unable to cancel one of the threads - so, with the original code, this test hangs for 10 seconds waiting for `timerAction` to complete before returning `Right 10`.